ggml.c

```diff
@@ -505,18 +505,29 @@ static inline __m256 sum_i16_pairs_float(const __m256i x) {
 
 // multiply int8_t, add results pairwise twice and return as float vector
 static inline __m256 mul_sum_i8_pairs_float(const __m256i x, const __m256i y) {
+#if __AVXVNNIINT8__
+    const __m256i zero = _mm256_setzero_si256();
+    const __m256i summed_pairs = _mm256_dpbssd_epi32(zero, x, y);
+    return _mm256_cvtepi32_ps(summed_pairs);
+#else
     // Get absolute values of x vectors
     const __m256i ax = _mm256_sign_epi8(x, x);
     // Sign the values of the y vectors
     const __m256i sy = _mm256_sign_epi8(y, x);
     // Perform multiplication and create 16-bit values
     const __m256i dot = _mm256_maddubs_epi16(ax, sy);
     return sum_i16_pairs_float(dot);
+#endif
 }
 
 static inline __m128i packNibbles( __m256i bytes )
 {
     // Move bits within 16-bit lanes from 0000_abcd_0000_efgh into 0000_0000_abcd_efgh
+#if __AVX512F__
+    const __m256i bytes_srli_4 = _mm256_srli_epi16(bytes, 4);   // 0000_0000_abcd_0000
+    bytes = _mm256_or_si256(bytes, bytes_srli_4);               // 0000_abcd_abcd_efgh
+    return _mm256_cvtepi16_epi8(bytes);                         // abcd_efgh
+#else
```

Comment:
This looks great, thank you! I made a [microbenchmark](https://github.com/dfyz/llama-avx-512/blob/master/pack_nibbles.cpp), and it shows a nice improvement on a Tiger Lake CPU:
```
-----------------------------------------------------------------
Benchmark                       Time             CPU   Iterations
-----------------------------------------------------------------
BenchPackNibblesAvx2          583 ns          582 ns      1212627
BenchPackNibblesAvx512        514 ns          513 ns      1313028
```

I don't think `packNibbles()` is on the hot path during inference, so it will probably *not* affect the overall inference speed (might be wrong here). However, it's makes the code both more readable and more performant, and is worth merging.